pkg/controller/atlascustomrole/atlascustomrole_controller.go

```diff
@@ -0,0 +1,195 @@
+	}
+	workflowCtx.SetConditionTrue(api.ResourceVersionStatus).SetConditionTrue(api.ValidationSucceeded)
+
+	if !r.AtlasProvider.IsResourceSupported(atlasCustomRole) {
+		return r.terminate(workflowCtx, atlasCustomRole,
+			api.ProjectCustomRolesReadyType, workflow.AtlasGovUnsupported,
+			false,
+			fmt.Errorf("the %T is not supported by Atlas for government", atlasCustomRole)), nil
+	}
+
+	atlasSdkClient, _, err := r.AtlasProvider.SdkClient(workflowCtx.Context,
+		&client.ObjectKey{
+			Namespace: atlasCustomRole.GetNamespace(),
+			Name:      atlasCustomRole.Spec.Credentials().Name,
+		},
+		workflowCtx.Log)
+	if err != nil {
+		return r.terminate(workflowCtx,
+			atlasCustomRole,
+			api.ProjectCustomRolesReadyType,
+			workflow.AtlasAPIAccessNotConfigured,
+			true,
+			fmt.Errorf("unable to create atlas client: %s", err.Error())), nil
+	}
+	workflowCtx.SdkClient = atlasSdkClient
+
+	if res := handleCustomRole(workflowCtx, atlasCustomRole, r.ObjectDeletionProtection); !res.IsOk() {
+		return r.fail(req, fmt.Errorf("%s", res.GetMessage())), nil
+	}
+	return r.idle(workflowCtx), nil
+}
+
+func (r *AtlasСustomRoleReconciler) terminate(
+	ctx *workflow.Context,
+	object akov2.AtlasCustomResource,
+	condition api.ConditionType,
+	reason workflow.ConditionReason,
+	retry bool,
+	err error,
+) ctrl.Result {
+	r.Log.Errorf("resource %T(%s/%s) failed on condition %s: %s", object, object.GetNamespace(), object.GetName(), condition, err)
+	result := workflow.Terminate(reason, err.Error())
+	ctx.SetConditionFromResult(condition, result)
+
+	if !retry {
+		result = result.WithoutRetry()
+	}
+
+	return result.ReconcileResult()
+}
+
+func (r *AtlasСustomRoleReconciler) idle(ctx *workflow.Context) ctrl.Result {
+	ctx.SetConditionTrue(api.ReadyType)
+	return workflow.OK().ReconcileResult()
+}
+
+// fail terminates the reconciliation silently(no updates on conditions)
+func (r *AtlasСustomRoleReconciler) fail(req ctrl.Request, err error) ctrl.Result {
+	r.Log.Errorf("Failed to query object %s: %s", req.NamespacedName, err)
+	return workflow.TerminateSilently().ReconcileResult()
+}
+
+// skip prevents the reconciliation to start and successfully return
+func (r *AtlasСustomRoleReconciler) skip() ctrl.Result {
+	r.Log.Infow(fmt.Sprintf("-> Skipping AtlasCustomRole reconciliation as annotation %s=%s",
+		customresource.ReconciliationPolicyAnnotation,
+		customresource.ReconciliationPolicySkip))
+	return workflow.OK().ReconcileResult()
+}
+
+// notFound terminates the reconciliation silently(no updates on conditions) and without retry
+func (r *AtlasСustomRoleReconciler) notFound(req ctrl.Request) ctrl.Result {
+	r.Log.Infof("Object %s doesn't exist, was it deleted after reconcile request?", req.NamespacedName)
+	return workflow.TerminateSilently().WithoutRetry().ReconcileResult()
+}
+
+func (r *AtlasСustomRoleReconciler) SetupWithManager(mgr ctrl.Manager, skipNameValidation bool) error {
```

Comment:
I believe this lacks indexers for when the credentials change (when using IDs). Eg:
https://github.com/mongodb/mongodb-atlas-kubernetes/blob/main/pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go#L223-L227

Reply:
Thanks, fixed